### PR TITLE
Moving review test to dashboard

### PR DESCRIPTION
**Pull request by vrackovic**
- `ReviewTest` is activated by pressing `Edit` icon in dashboard.
- It is opened as modal
- Component is now loading real `Test` data from backend

### CODE REVIEW

File: web/src/main/frontend/src/base/stores/CompletedTestStore.js [hunk 1]
```diff
@@ -1,74 +1,9 @@
...
-      type: 'CodeChallenge',
-      question: 'Hello fellow coder, please write us some clean code here:',
-      solution:
-        'if (true) {// give me points please, I need this job to support my family \n }',
-      availablePoints: 40
-    }
-  ];
+  @observable test;
 
+  @observable activateReview = false;
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CompletedTestStore.js]: Rename to `reviewActive`.

File: web/src/main/frontend/src/components/Test.js [hunk 2]
```diff
@@ -59,7 +59,8 @@ class Test extends React.Component {
         <div className="review__container__header row-eq-height">
           <div className="review__container__counter row-eq-height d-flex justify-content-center ">
             <span>
-              Step {this.counter || 1} of {this.props.getQuestions().length}
+              Step {this.counter || 1} of
+              {this.props.questions && this.props.questions.length}
```
Comment (iscekic) [inline web/src/main/frontend/src/components/Test.js]: When is `questions` falsy?

File: web/src/main/frontend/src/pages/TakeTest.js [hunk 3]
```diff
@@ -13,7 +13,9 @@ class TakeTest extends React.Component {
   }
 
   questionsToReview = () =>
-    this.props.newTest.test.questions.map(getComponentByQuestionType);
+    this.props.newTest.test.questions.map(question =>
```
Comment (iscekic) [inline web/src/main/frontend/src/pages/TakeTest.js]: Equivalent:

```js
this.props.newTest.test.questions.map(getComponentByQuestionType)
```

File: web/src/main/frontend/src/components/ReviewTest.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import CSSTransition from 'react-transition-group/CSSTransition';
+import {inject, observer} from 'mobx-react';
+import Test from './Test';
+import getComponentByQuestionType from './utility/getComponentByQuestionType';
+
+@inject('confirmPopup', 'completedTest')
+@observer
+class ReviewTest extends React.Component {
+  render() {
+    const questionsToReview = this.props.completedTest.test.questions.map(
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewTest.js]: Can this be computed?

File: web/src/main/frontend/src/components/ReviewTest.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      question => getComponentByQuestionType(question, true)
+    );
+
+    return (
+      <div>
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewTest.js]: Do we need to wrap with `div`?

File: web/src/main/frontend/src/components/Test.js (hunk 2 above)
Comment (vrackovic) [reply web/src/main/frontend/src/components/Test.js]: This was an issue when function has been triggered before data was loaded. It is now fixed and `questions` will never be falsy so check will be removed.

File: web/src/main/frontend/src/components/ReviewTest.js (hunk 5 above)
Comment (vrackovic) [reply web/src/main/frontend/src/components/ReviewTest.js]: Nope, we don't

File: web/src/main/frontend/src/base/stores/CompletedTestStore.js [hunk 6]
```diff
@@ -83,6 +18,22 @@ class CompletedTestStore {
...
+    this.reviewActive = true;
+  }
+
+  @action.bound
+  exitReview() {
+    this.reviewActive = false;
+  }
+
+  @computed
+  get testQuestions() {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CompletedTestStore.js]: Will there be some computation here in the future?

File: web/src/main/frontend/src/components/CodeChallengeQuestion.js [hunk 7]
```diff
@@ -61,7 +61,7 @@ class CodeChallengeQuestion extends Component {
             value={this.state.value}
             onBeforeChange={this.updateCode}
             options={this.getOptions()}
-            mode={this.options.mode}
+            mode={this.state.mode}
```
Comment (iscekic) [inline web/src/main/frontend/src/components/CodeChallengeQuestion.js]: `state.mode`? Where is this defined?

File: web/src/main/frontend/src/components/ReviewTest.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import CSSTransition from 'react-transition-group/CSSTransition';
+import {inject, observer} from 'mobx-react';
+import Test from './Test';
+import getComponentByQuestionType from './utility/getComponentByQuestionType';
+
+@inject('confirmPopup', 'completedTest')
+@observer
+class ReviewTest extends React.Component {
+  render() {
+    const questionsToReview = this.props.completedTest.testQuestions.map(
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewTest.js]: Can this be `computed`?

Comment (vrackovic) [reply web/src/main/frontend/src/components/ReviewTest.js]: I can map trough `testQuestions` and return `question => getComponentByQuestionType(question, true)` from store, but question is is it smart to move jsx elements as computed in store? `getComponentByQuestionType` returns element render that depends of `question.type`

File: web/src/main/frontend/src/components/CodeChallengeQuestion.js (hunk 7 above)
Comment (vrackovic) [reply web/src/main/frontend/src/components/CodeChallengeQuestion.js]: This field should say `this.state.syntax`. It was called `mode` before renamed to `syntax`.

File: web/src/main/frontend/src/components/ReviewTest.js (hunk 8 above)
Comment (iscekic) [reply web/src/main/frontend/src/components/ReviewTest.js]: Hm, you're right. Let's leave this here then.

File: web/src/main/frontend/src/base/stores/CompletedTestStore.js (hunk 6 above)
Comment (vrackovic) [reply web/src/main/frontend/src/base/stores/CompletedTestStore.js]: Same as comment about returning `JSX` from store, that is only place that calls `testQuestion`

**Pull request merged**